daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -283,6 +283,13 @@ sealed abstract class HasTxNodes[Nid, +Cid, +Val] {
 
   def roots: ImmArray[Nid]
 
+  def byKeyNodes: Set[Nid] =
+    nodes.collect {
+      case (nodeId, node: Node.NodeExercises[_, _, _]) if node.byKey.getOrElse(false) => nodeId
```

Comment:
Does this really do what we want? Or do we need
```suggestion
      case (nodeId, node): (_, Node.NodeExercises[_, _, _]) if node.byKey.getOrElse(false) => nodeId
```
There is [this related forum post](https://discuss.daml.com/t/v-ty-in-a-pattern-is-not-an-annotation-scala-ritual/1450) from @S11001001, which really scares me.

Reply:
We are actually doing pattern matching, so it is fine to use my initial code. Concerning @S11001001's I will argue you should avoid using pattern matching for initializing value. 

Your code is not working anyway, because of type erasure, i.e. the type `(_, Node.NodeExercises[_, _, _])` becomes at compile time `(_, _)`. 

By tricking a bit you proposal I manage to please the compiler with:
```suggestion
       case tuple: (_, Node.NodeExercises[_, _, _]) if tuple._1.byKey.getOrElse(false) => tuple._2
```
Note it however emit the "nice" warning: 
```
.../Transaction.scala:288: warning: non-variable type argument com.daml.lf.transaction.Node.NodeExercises[_,_,_] in type pattern (_, com.daml.lf.transaction.Node.NodeExercises[_,_,_]) is unchecked since it is eliminated by erasure
      case tuple: (_, Node.NodeExercises[_, _, _]) if tuple._1.byKey.getOrElse(false) => tuple._2
```

I think this should be an error not a warning.

Reply:
This is so tricky. Thanks for your explanation.

Reply:
The prior code is the proper way to express this, because indeed a pattern is what you want here. The advice about placement to get an annotation from my post doesn't apply here, because there is nowhere you can put an annotation in a `case` anyway.